File: 0992-subarrays-with-k-different-integers/0992-subarrays-with-k-different-integers.cpp
class Solution {
public:
    int calculateGoodSubarraysTillK(vector<int>&nums, int k){
        int n = nums.size(), ans = 0;
        int i = 0, j = 0;
        unordered_map<int, int>mp;
        
        while(j < n){
            ++mp[nums[j]];
            while(mp.size() > k){
                --mp[nums[i]];
                if(mp[nums[i]] == 0){
                    mp.erase(nums[i]);
                }
                ++i;
            }
            ans = ans + (j - i + 1);
            ++j;
        }
        return ans;
    }
    int subarraysWithKDistinct(vector<int>& nums, int k) {        
        int tillK = calculateGoodSubarraysTillK(nums, k);
        int tillK_1 = calculateGoodSubarraysTillK(nums, k - 1);
        return tillK - tillK_1;
    }
};

File: 0992-subarrays-with-k-different-integers/0992-subarrays-with-k-different-integers.java
class Solution {
    public int helper(int[] nums, int k){
        int n = nums.length;
        int i = 0, j = 0;
        int res = 0;
        
        Map<Integer, Integer>map = new HashMap<>();
        
        while(j < n){
            map.put(nums[j], map.getOrDefault(nums[j], 0) + 1);
            
            while(map.size() > k){
                map.put(nums[i], map.get(nums[i]) - 1);
                
                if(map.get(nums[i]) == 0)   
                    map.remove(nums[i]);
                
                ++i;
            }
            
            res += (j - i + 1);
            ++j;
        }
        
        return res;        
    }
    public int subarraysWithKDistinct(int[] nums, int k) {
        int withK = helper(nums, k);
        int withOutK = helper(nums, k - 1);
        
        return withK - withOutK;
    }
}